File: grooves-example-javaee/src/main/java/grooves/example/javaee/domain/PatientAccount.java
package grooves.example.javaee.domain;

import static rx.Observable.*;
import static rx.RxReactiveStreams.toPublisher;

import com.github.rahulsom.grooves.api.snapshots.Snapshot;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import javax.xml.bind.annotation.XmlTransient;
import lombok.Data;
import org.jetbrains.annotations.NotNull;
import org.reactivestreams.Publisher;

// tag::documented[]
@Data
public class PatientAccount
        implements Snapshot<Patient, Long, Long, PatientEvent>, // <1>
                Serializable {
    private Long id;
    private Patient aggregate;
    private Patient deprecatedBy;
    private List<Patient> deprecates = new ArrayList<>();
    private long lastEventPosition; // <2>
    private Date lastEventTimestamp; // <3>

    private String name;
    private BigDecimal balance = new BigDecimal(0);
    private BigDecimal moneyMade = new BigDecimal(0);

    // end::documented[]

    @NotNull
    @Override
    @XmlTransient
    // tag::documented[]
    public Publisher<Patient> getAggregateObservable() { // <4>
        return toPublisher(aggregate != null ? just(aggregate) : empty());
    }

    // end::documented[]
    @NotNull
    @Override
    @XmlTransient
    // tag::documented[]
    public Publisher<Patient> getDeprecatedByObservable() { // <5>
        return toPublisher(deprecatedBy != null ? just(deprecatedBy) : empty());
    }

    // end::documented[]
    @NotNull
    @Override
    @XmlTransient
    // tag::documented[]
    public Publisher<Patient> getDeprecatesObservable() { // <6>
        return toPublisher(from(deprecates));
    }
    // end::documented[]

    @Override
    public String toString() {
        return String.format(
                "PatientAccount{id=%d, aggregate=%s, lastEventPosition=%d, lastEventTimestamp=%s}",
                id, aggregate, lastEventPosition, lastEventTimestamp);
    }

    // tag::documented[]
}
// end::documented[]

File: grooves-example-springboot-kotlin/src/main/kotlin/grooves/boot/kotlin/domain/PatientAccount.kt
package grooves.boot.kotlin.domain

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonIgnore
import com.github.rahulsom.grooves.api.snapshots.Snapshot
import grooves.boot.kotlin.repositories.PatientRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Configurable
import reactor.core.publisher.Flux.empty
import reactor.core.publisher.Mono
import reactor.core.publisher.Mono.just
import java.math.BigDecimal
import java.util.Date

// tag::documented[]
@Configurable
class PatientAccount : Snapshot<Patient, String, String, PatientEvent> { // <1>

    override var id: String? = null
    override var lastEventPosition: Long = 0 // <2>

    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ")
    override var lastEventTimestamp: Date? = null // <3>
    val deprecatesIds: MutableList<String> = mutableListOf()
    private var deprecator: Patient? = null
    var aggregateId: String? = null

    var name: String? = null
    var balance: BigDecimal = BigDecimal.ZERO
    var moneyMade: BigDecimal = BigDecimal.ZERO

    fun getDeprecatedBy() = deprecator

    @JsonIgnore
    override fun getAggregateObservable() =
        // <4>
        aggregateId?.let { patientRepository!!.findAllById(just(it)) } ?: empty()

    override fun setAggregate(aggregate: Patient) {
        this.aggregateId = aggregate.id
    }

    @JsonIgnore
    @Autowired
    var patientRepository: PatientRepository? = null

    @JsonIgnore
    override fun getDeprecatedByObservable() =
        // <5>
        deprecator?.let { just(it) } ?: Mono.empty()

    override fun setDeprecatedBy(deprecatingAggregate: Patient) {
        deprecator = deprecatingAggregate
    }

    @JsonIgnore
    override fun getDeprecatesObservable() =
        // <6>
        if (deprecatesIds.isEmpty()) {
            empty()
        } else {
            patientRepository!!.findAllById(deprecatesIds)
        }

    // end::documented[]
    override fun toString(): String =
        "PatientAccount(id=$id, aggregate=$aggregateId, " +
            "lastEventPosition=$lastEventPosition, lastEventTimestamp=$lastEventTimestamp)"
    // tag::documented[]
}
// end::documented[]